<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title> Multivariate Analysis </title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Multivariate Analysis_files/libs/clipboard/clipboard.min.js"></script>
<script src="Multivariate Analysis_files/libs/quarto-html/quarto.js"></script>
<script src="Multivariate Analysis_files/libs/quarto-html/popper.min.js"></script>
<script src="Multivariate Analysis_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Multivariate Analysis_files/libs/quarto-html/anchor.min.js"></script>
<link href="Multivariate Analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Multivariate Analysis_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Multivariate Analysis_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Multivariate Analysis_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Multivariate Analysis_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


<link rel="stylesheet" href="assets/rany_style.css">
</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Outline</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span style="color: #234F1E;"><strong>Introduction</strong></span></a></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering"><span style="color: #234F1E;"><strong>Clustering</strong></span></a>
  <ul class="collapse">
  <li><a href="#partitioningcentroid-based-clustering" id="toc-partitioningcentroid-based-clustering" class="nav-link" data-scroll-target="#partitioningcentroid-based-clustering"><span style="color: #002D62;">1. Partitioning/Centroid based clustering</span></a></li>
  <li><a href="#hierarchicalagglomerative-clustering" id="toc-hierarchicalagglomerative-clustering" class="nav-link" data-scroll-target="#hierarchicalagglomerative-clustering"><span style="color: #002D62;">2. Hierarchical(Agglomerative) clustering</span></a></li>
  <li><a href="#grid-based-clustering" id="toc-grid-based-clustering" class="nav-link" data-scroll-target="#grid-based-clustering"><span style="color: #002D62;">3. Grid-Based clustering</span></a></li>
  <li><a href="#density-based-clustering" id="toc-density-based-clustering" class="nav-link" data-scroll-target="#density-based-clustering"><span style="color: #002D62;">4. Density-Based clustering</span></a></li>
  <li><a href="#model-based-clustering" id="toc-model-based-clustering" class="nav-link" data-scroll-target="#model-based-clustering"><span style="color: #002D62;">5. Model-Based clustering</span></a></li>
  <li><a href="#example" id="toc-example" class="nav-link" data-scroll-target="#example"><span style="color: #002D62;">Example</span></a></li>
  </ul></li>
  <li><a href="#k-means-clustering" id="toc-k-means-clustering" class="nav-link" data-scroll-target="#k-means-clustering"><span style="color: #234F1E;"><strong>K-Means Clustering</strong></span></a>
  <ul class="collapse">
  <li><a href="#k-means-clustering-steps" id="toc-k-means-clustering-steps" class="nav-link" data-scroll-target="#k-means-clustering-steps"><span style="color: #002D62;">K-Means Clustering Steps</span></a></li>
  <li><a href="#example-1" id="toc-example-1" class="nav-link" data-scroll-target="#example-1"><span style="color: #002D62;">Example</span></a></li>
  <li><a href="#k-means-clustering-visualization" id="toc-k-means-clustering-visualization" class="nav-link" data-scroll-target="#k-means-clustering-visualization"><span style="color: #002D62;">K-Means Clustering Visualization</span></a></li>
  </ul></li>
  <li><a href="#principal-component-analysis-pca" id="toc-principal-component-analysis-pca" class="nav-link" data-scroll-target="#principal-component-analysis-pca"><span style="color: #234F1E;"><strong>Principal Component Analysis (PCA)</strong></span></a>
  <ul class="collapse">
  <li><a href="#heres-how-pca-typically-works" id="toc-heres-how-pca-typically-works" class="nav-link" data-scroll-target="#heres-how-pca-typically-works"><span style="color: #002D62;">Here’s how PCA typically works</span></a></li>
  <li><a href="#application-of-pca" id="toc-application-of-pca" class="nav-link" data-scroll-target="#application-of-pca"><span style="color: #002D62;">Application of PCA</span></a></li>
  <li><a href="#example-2" id="toc-example-2" class="nav-link" data-scroll-target="#example-2"><span style="color: #002D62;">Example</span></a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><p style="color:black,text-align:center"><br>
Multivariate Analysis<br>
</p></h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author"><font color="#ff6600"><b>Biometrics Unit</b></font> </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <font color="#ff6600"><b>International Institute of Tropical Agriculture (IITA)</b></font>
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  


</header>


<section id="introduction" class="level1">
<h1><span style="color: #234F1E;"><strong>Introduction</strong></span></h1>
<p>Data analysis is a vital part of every research work. To achieve this, there are three types of analysis that can be appliedto datasets, these includes:</p>
<ol type="1">
<li>Univariate analysis: This involves only <strong>one</strong> variable (methods includes mode, mean, median, etc.).</li>
<li>Bivariate analysis: This involves <strong>two</strong> variables (method includes linear regression, correlation etc. )</li>
<li>Multivariate analysis: This involves <strong>two or more</strong> variables ( Cluster analysis, Principal Components or Factor analysis, neural network Bayesian classifier, matrix plot etc.). In multivariate analysis we are concerned with sets of objects on each of which p variables (or variates) are measured, but usually with no prior differentiation of variables into causes and effects. We can look at any one variable in isolation but to get the whole picture the variables must be considered jointly.</li>
</ol>
<p>Multivariate analysis(MVA) refers to a broad set of statistical methods designed for examining relationships among multiple variables concurrently, typically beyond two. Its purpose is to uncover intricate patterns and correlations within a dataset, offering a richer and more refined comprehension of the underlying scenario compared to simpler analyses. Through simultaneous examination of multiple variables, MVA yields deeper insights and more precise predictions, thereby bolstering decision-making within data-driven industries.</p>
<p>In fields like plant breeding, where understanding complex interconnections among data is crucial, this technique plays a foundation role. Variables such as days to flowering, days to maturity, 100_seed_weights and plant heights collectively influence the yield of crops. This necessitates the need for multivariate analysis.</p>
<p>There are several methods of multivariate analysis, but in this study, we will be considering three major methods:</p>
<ul>
<li>Clustering</li>
<li>K-means clustering, and</li>
<li>Principal Component Analysis (PCA).</li>
</ul>
</section>
<section id="clustering" class="level1">
<h1><span style="color: #234F1E;"><strong>Clustering</strong></span></h1>
<p>A cluster analysis groups observations or variables based on similarities between them. When organizing data into clusters, the objective is for variables within the same cluster to exhibit greater similarity to each other compared to variables in different clusters. This comparison is quantified through intracluster and intercluster distances. Intracluster distance assesses the proximity of data points within a single cluster, ideally aiming for a minimal value. Intercluster distance, on the other hand, measures the separation between data points belonging to distinct clusters, with the goal of maximizing this distance for effective clustering. Data clustering techniques are valuable tools for researchers working with large databases of multivariate data.</p>
<p>Clustering analysis aims to uncover patterns within data and organize them into distinct groups based on those patterns. Hence, if two data points share similar traits, it indicates they follow the same pattern and thus should be grouped together. Through clustering analysis, we can explore which features tend to co-occur and understand the defining characteristics of each group.</p>
<p>We classify or cluster:</p>
<ol type="a">
<li>For data description - to understand the data better than producing a distribution.</li>
<li>For data reduction – i.e.&nbsp;reduce the size of data for, say, geographical reasons, etc.</li>
<li>For typology/taxonomy - to look for some natural grouping.</li>
</ol>
<p><strong>The major type of clustering includes:</strong></p>
<ol type="1">
<li>Partitioning/Centroid based clustering</li>
<li>Hierarchical clustering</li>
<li>Grid-Based clustering</li>
<li>Density-Based clustering</li>
<li>Model-Based clustering</li>
</ol>
<section id="partitioningcentroid-based-clustering" class="level2">
<h2 class="anchored" data-anchor-id="partitioningcentroid-based-clustering"><span style="color: #002D62;">1. Partitioning/Centroid based clustering</span></h2>
<p>Centroid-based clustering is a clustering approach that divides a dataset into comparable groups by assessing the distances between their centroids. Example, k-means clustering.</p>
<p><img src="images/centroid-based.png" width="65%"></p>
</section>
<section id="hierarchicalagglomerative-clustering" class="level2">
<h2 class="anchored" data-anchor-id="hierarchicalagglomerative-clustering"><span style="color: #002D62;">2. Hierarchical(Agglomerative) clustering</span></h2>
<p>This method identifies clusters by evaluating the proximity of data points to each other, operating on the premise that objects in closer proximity are more closely associated than those situated farther apart. To apply connectivity-based clustering, you must first select the relevant data points and quantify their similarities or differences using a distance metric. Next, a connectivity measure, like a graph or network, is created to depict the connections among the data points. Subsequently, the clustering algorithm utilizes this connectivity data to organize the data points into clusters that capture their inherent similarities. This process is often represented visually through a dendrogram, resembling a hierarchical tree structure.</p>
<p>The Agglomerative clustering is the most common type of hierarchical clustering used to group objects in clusters based on their similarity. It’s also known as AGNES (Agglomerative Nesting). The algorithm starts by treating each object as a singleton cluster. Next, pairs of clusters are successively merged until all clusters have been merged into one big cluster containing all objects. The result is a tree-based representation of the objects, named dendrogram.</p>
<p>A <strong>dendrogram</strong> shows the hierarchical relationship between the clusters</p>
<p><img src="images/dendo.png" width="65%"></p>
<p>There are many distance metrics and the choice depends on the type of data:</p>
<ul>
<li>If the data are continuous quantitative, we can use the Euclidean distance or that of Manhattan,</li>
<li>If the data is binary (categorical), we can use the Jaccard distance</li>
<li>Other distance measurements include Minkowski, Canberra, etc.</li>
</ul>
<p>Where there is no theoretical justification for an alternative, the <strong>Euclidean distance</strong> should generally be preferred. After the distance metric, we then have to choose a linkage criteria. There are many options: single-linkage, complete-linkage, mean or average-linkage, Ward’s method, etc. Each of the methods will produce a different dendrogram. In practice, we will most often prefer the <strong>Ward method</strong>. Ward’s method seeks to minimize intra-class variability and to maximize inter-class variability to obtain the most homogeneous possible clusters</p>
<p><strong>Important things to note when working with Hierarchical clustering:</strong></p>
<ol type="1">
<li><p><strong>Standardize the data:</strong> The variables to be used for clustering are of different units. Standardizing the data (mean zero, unit variance) will ensure that the data is at the same scale. We subtract each data from its mean and divide it by the standard deviation. We can use the <code>scale ()</code> function in R</p></li>
<li><p><strong>Dealing with missing values:</strong> Several ways to deal with these values,</p>
<ul>
<li>delete them</li>
<li>impute them with a mean, median, mode or use advanced regression techniques.</li>
</ul></li>
</ol>
<p>The root of the dendrogram corresponds to the cluster with all the individuals together. This dendrogram represents a hierarchy of partitions. The “fusion” distance (<strong>height</strong>) is indicated on the <code>y-axis</code> of the dendrogram</p>
</section>
<section id="grid-based-clustering" class="level2">
<h2 class="anchored" data-anchor-id="grid-based-clustering"><span style="color: #002D62;">3. Grid-Based clustering</span></h2>
<p>Grid-based clustering divides a high-dimensional dataset into cells, which are distinct sets of non-overlapping sub-regions. Every cell is given a unique identifier known as a cell ID, and all data points contained within a cell are regarded as belonging to the same cluster. This clustering technique proves efficient for analyzing expansive multidimensional datasets, as it diminishes the search time required for locating nearest neighbors, a frequent procedure in numerous clustering methodologies.</p>
<p><img src="images/grid-based.png" width="65%"></p>
</section>
<section id="density-based-clustering" class="level2">
<h2 class="anchored" data-anchor-id="density-based-clustering"><span style="color: #002D62;">4. Density-Based clustering</span></h2>
<p>Density-based clustering represents a robust unsupervised machine learning method that enables the identification of dense clusters within a dataset. In contrast to clustering algorithms like K-means and hierarchical clustering, density-based clustering can unveil clusters of varied shapes, sizes, and densities. This technique proves particularly advantageous when handling datasets containing noise or outliers, or when the precise number of clusters within the data is unknown.</p>
<p><img src="images/density.png" width="65%"></p>
</section>
<section id="model-based-clustering" class="level2">
<h2 class="anchored" data-anchor-id="model-based-clustering"><span style="color: #002D62;">5. Model-Based clustering</span></h2>
<p>Model-based clustering organizes data points into groups according to their probability distribution, leveraging statistical patterns to delineate clusters within the dataset. Unlike centroid-based clustering, it employs statistical patterns for cluster identification. However, model-based clustering is susceptible to overfitting, wherein clustering excessively relies on the dataset, leading to inaccurate predictions.</p>
<p><img src="images/model.png" width="65%"></p>
</section>
<section id="example" class="level2">
<h2 class="anchored" data-anchor-id="example"><span style="color: #002D62;">Example</span></h2>
<p>Let’s Consider the <code>steptoe.morex.pheno.csv</code> data from the <code>agridat</code> package, multi-environment trial of barley. The dataset contains 2432 observations on the following 10 variables:</p>
<ul>
<li><em>gen</em> (genotype) - a Character vector for the genes of the barley</li>
<li><em>env</em> (environment)- a character vector of the environment</li>
<li><em>amylase</em> (alpha amylase), 20 Deg Units - a numeric vector for the alpha amylase sample</li>
<li><em>diapow</em> (diastatic power), degree units - a numeric vector for the diastatic power</li>
<li><em>hddate</em> (heading date),julian days - a numeric vector for the heading date</li>
<li><em>lodging</em>, percent- a numeric vector for the lodging</li>
<li><em>malt</em> (malt extract), percent - numeric vector for the malt sample</li>
<li><em>height</em> (plant height), centimeters - numeric vector for the height sample</li>
<li><em>protein</em> (grain protein), percent - numeric vector for the protein sample</li>
<li><em>yield</em> (grain yield), mt / ha - numeric vector for the plant yield</li>
</ul>
<p>Lets take a look at a sample of the data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"steptoe.morex.pheno.csv"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>dat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,432 × 10
   gen     env   amylase diapow hddate lodging  malt height protein yield
   &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
 1 Steptoe MN92     22.7     46   150.      NA  73.6   84.5    10.5  5.53
 2 Steptoe MTi92    30.1     72   178       10  76.5   NA      11.2  8.64
 3 Steptoe MTd92    26.7     78   165       15  74.5   75.5    13.4  5.90
 4 Steptoe ID91     26.2     74   179       NA  74.1  111      12.1  8.63
 5 Steptoe OR91     19.6     62   191       NA  71.5   90      11.7  5.34
 6 Steptoe WA91     23.6     54   181       NA  73.8  112      10    6.27
 7 Steptoe MTi91    21       62   181       NA  70.8   98      12    4.10
 8 Steptoe MTd91    NA       NA   181       NA  NA     82      NA    7.07
 9 Steptoe NY92     NA       NA   176        0  NA     77.5    NA    6.05
10 Steptoe ON92     NA       NA   198       50  NA     95      NA    3.70
# ℹ 2,422 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(dat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>spc_tbl_ [2,432 × 10] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ gen    : chr [1:2432] "Steptoe" "Steptoe" "Steptoe" "Steptoe" ...
 $ env    : chr [1:2432] "MN92" "MTi92" "MTd92" "ID91" ...
 $ amylase: num [1:2432] 22.7 30.1 26.7 26.2 19.6 23.6 21 NA NA NA ...
 $ diapow : num [1:2432] 46 72 78 74 62 54 62 NA NA NA ...
 $ hddate : num [1:2432] 150 178 165 179 191 ...
 $ lodging: num [1:2432] NA 10 15 NA NA NA NA NA 0 50 ...
 $ malt   : num [1:2432] 73.6 76.5 74.5 74.1 71.5 73.8 70.8 NA NA NA ...
 $ height : num [1:2432] 84.5 NA 75.5 111 90 112 98 82 77.5 95 ...
 $ protein: num [1:2432] 10.5 11.2 13.4 12.1 11.7 10 12 NA NA NA ...
 $ yield  : num [1:2432] 5.53 8.64 5.9 8.63 5.34 ...
 - attr(*, "spec")=
  .. cols(
  ..   gen = col_character(),
  ..   env = col_character(),
  ..   amylase = col_double(),
  ..   diapow = col_double(),
  ..   hddate = col_double(),
  ..   lodging = col_double(),
  ..   malt = col_double(),
  ..   height = col_double(),
  ..   protein = col_double(),
  ..   yield = col_double()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
</div>
<p>This shows us the structure and different data types in our dataset</p>
<p>Let’s look at the summary of the data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(dat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     gen                env               amylase          diapow      
 Length:2432        Length:2432        Min.   :14.90   Min.   : 35.00  
 Class :character   Class :character   1st Qu.:25.62   1st Qu.: 70.00  
 Mode  :character   Mode  :character   Median :28.50   Median : 84.00  
                                       Mean   :29.04   Mean   : 87.25  
                                       3rd Qu.:32.00   3rd Qu.:101.00  
                                       Max.   :49.30   Max.   :229.00  
                                       NA's   :1066    NA's   :1066    
     hddate         lodging            malt           height      
 Min.   :143.0   Min.   :  0.00   Min.   :69.00   Min.   : 34.00  
 1st Qu.:174.5   1st Qu.: 15.00   1st Qu.:73.30   1st Qu.: 82.50  
 Median :183.5   Median : 35.00   Median :74.50   Median : 95.00  
 Mean   :181.2   Mean   : 37.13   Mean   :74.72   Mean   : 94.95  
 3rd Qu.:191.0   3rd Qu.: 55.00   3rd Qu.:75.90   3rd Qu.:109.00  
 Max.   :217.0   Max.   :100.00   Max.   :83.00   Max.   :151.00  
                 NA's   :1521     NA's   :1067    NA's   :5       
    protein          yield       
 Min.   : 8.00   Min.   : 1.390  
 1st Qu.:11.90   1st Qu.: 4.092  
 Median :13.00   Median : 5.272  
 Mean   :12.92   Mean   : 5.293  
 3rd Qu.:14.00   3rd Qu.: 6.338  
 Max.   :17.50   Max.   :11.526  
 NA's   :1067                    </code></pre>
</div>
</div>
<p>We have data from 17 environments</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(dat<span class="sc">$</span>env)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>NULL</code></pre>
</div>
</div>
<p>Let’s consider only the env <strong>ID91</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>dat.ID91 <span class="ot">&lt;-</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(env<span class="sc">==</span><span class="st">"ID91"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We take a look at the summary of the data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(dat.ID91)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     gen                env               amylase          diapow      
 Length:152         Length:152         Min.   :19.10   Min.   : 47.00  
 Class :character   Class :character   1st Qu.:25.48   1st Qu.: 72.00  
 Mode  :character   Mode  :character   Median :28.05   Median : 83.00  
                                       Mean   :28.08   Mean   : 84.16  
                                       3rd Qu.:30.12   3rd Qu.: 94.00  
                                       Max.   :40.10   Max.   :144.00  
                                                                       
     hddate         lodging         malt           height         protein     
 Min.   :173.0   Min.   : NA   Min.   :71.10   Min.   : 92.7   Min.   :11.30  
 1st Qu.:178.0   1st Qu.: NA   1st Qu.:73.40   1st Qu.:108.9   1st Qu.:12.90  
 Median :180.5   Median : NA   Median :74.40   Median :114.3   Median :13.50  
 Mean   :180.5   Mean   :NaN   Mean   :74.36   Mean   :114.0   Mean   :13.57  
 3rd Qu.:183.0   3rd Qu.: NA   3rd Qu.:75.20   3rd Qu.:119.4   3rd Qu.:14.22  
 Max.   :188.0   Max.   : NA   Max.   :78.50   Max.   :137.2   Max.   :16.00  
                 NA's   :152                                                  
     yield       
 Min.   : 4.048  
 1st Qu.: 6.681  
 Median : 7.684  
 Mean   : 7.500  
 3rd Qu.: 8.382  
 Max.   :10.315  
                 </code></pre>
</div>
</div>
<p><code>lodging</code> is missing for all genotypes in this environment. So we exclude <code>lodging</code> in the dataset</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>dat.ID91 <span class="ot">&lt;-</span> dat.ID91 <span class="sc">%&gt;%</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>lodging)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dat.ID91)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "gen"     "env"     "amylase" "diapow"  "hddate"  "malt"    "height" 
[8] "protein" "yield"  </code></pre>
</div>
</div>
<p>No more missing data, but it’s a good practice to use the function <code>na.omit()</code> and exclude any missing data the dataset may have</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>dat.ID91 <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(dat.ID91)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s store the names of the genotypes in a vector <code>dat.ID91.label</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>dat.ID91.label <span class="ot">&lt;-</span> dat.ID91<span class="sc">$</span>gen</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We delete the gen and env columns from the dataset to only keep the quantitative variables</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>dat.ID91<span class="sc">$</span>gen <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>dat.ID91<span class="sc">$</span>env <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>dat.ID91</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 152 × 7
   amylase diapow hddate  malt height protein yield
     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
 1    26.2     74   179   74.1   111     12.1  8.63
 2    36.2     97   180   76.7   116     14.8  7.95
 3    32.8     83   183   74.4   117.    14.3  6.10
 4    24.9     81   178.  72.3   113     14    7.20
 5    30.2     69   180   72.9   117.    13.8  6.01
 6    30.3     99   184.  73.2   119.    14.9  8.09
 7    33.4     63   178   73.5   104.    13.1  7.98
 8    25.5     74   186.  71.8   130.    14.6  5.93
 9    26.8     78   185   75     124.    12.8  8.72
10    28.3     85   179   71.3   119.    14.1  6.48
# ℹ 142 more rows</code></pre>
</div>
</div>
<p>The data have different units. We standardize them using the function <code>scale()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>dat.ID91.sc <span class="ot">&lt;-</span> <span class="fu">scale</span>(dat.ID91)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(dat.ID91.sc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    amylase              diapow             hddate              malt         
 Min.   :-2.323173   Min.   :-2.13347   Min.   :-2.13124   Min.   :-2.39465  
 1st Qu.:-0.674579   1st Qu.:-0.69806   1st Qu.:-0.71724   1st Qu.:-0.70544  
 Median :-0.008677   Median :-0.06648   Median :-0.01023   Median : 0.02899  
 Mean   : 0.000000   Mean   : 0.00000   Mean   : 0.00000   Mean   : 0.00000  
 3rd Qu.: 0.527924   3rd Qu.: 0.56510   3rd Qu.: 0.69677   3rd Qu.: 0.61654  
 Max.   : 3.107488   Max.   : 3.43592   Max.   : 2.11078   Max.   : 3.04018  
     height            protein             yield        
 Min.   :-2.38931   Min.   :-2.44921   Min.   :-2.7274  
 1st Qu.:-0.57607   1st Qu.:-0.72269   1st Qu.:-0.6471  
 Median : 0.02835   Median :-0.07525   Median : 0.1461  
 Mean   : 0.00000   Mean   : 0.00000   Mean   : 0.0000  
 3rd Qu.: 0.59919   3rd Qu.: 0.70708   3rd Qu.: 0.6974  
 Max.   : 2.59152   Max.   : 2.62243   Max.   : 2.2246  </code></pre>
</div>
</div>
<p>The distance metric is calculated using the Euclidean distance</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>dist.ID91 <span class="ot">&lt;-</span> <span class="fu">dist</span>(dat.ID91.sc, <span class="at">method =</span> <span class="st">'euclidean'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We perform the hierarchical classification with the <code>hclust ()</code> function and specify the method</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>hclust.ID91 <span class="ot">&lt;-</span> <span class="fu">hclust</span>(dist.ID91, <span class="at">method =</span> <span class="st">'ward.D2'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can visualize the dendrogram using the <code>plot()</code> function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hclust.ID91, <span class="at">hang =</span> <span class="sc">-</span><span class="dv">1</span>,<span class="at">rect=</span> <span class="cn">TRUE</span>, <span class="at">cex =</span> <span class="fl">0.5</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Multivariate-Analysis_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can visualize the three clusters with different colors by using the <code>color_branches ()</code> function of the <code>dendextend</code> package</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dendextend)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>dend.ID91 <span class="ot">&lt;-</span> <span class="fu">as.dendrogram</span>(hclust.ID91)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>col.dend.ID91 <span class="ot">&lt;-</span> <span class="fu">color_branches</span>(dend.ID91, <span class="at">k =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(col.dend.ID91, <span class="at">hang =</span> <span class="dv">1</span>, <span class="at">cex =</span> <span class="fl">0.4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Multivariate-Analysis_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can get the groups to which the genotypes belong by specifying the number of clusters</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>cut.ID91 <span class="ot">&lt;-</span> <span class="fu">cutree</span>(hclust.ID91, <span class="at">k =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We add the group names of the genotypes to the original data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>dat.ID91 <span class="ot">&lt;-</span> dat.ID91 <span class="sc">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cluster =</span> cut.ID91)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>dat.ID91</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 152 × 8
   amylase diapow hddate  malt height protein yield cluster
     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;int&gt;
 1    26.2     74   179   74.1   111     12.1  8.63       1
 2    36.2     97   180   76.7   116     14.8  7.95       2
 3    32.8     83   183   74.4   117.    14.3  6.10       1
 4    24.9     81   178.  72.3   113     14    7.20       1
 5    30.2     69   180   72.9   117.    13.8  6.01       1
 6    30.3     99   184.  73.2   119.    14.9  8.09       3
 7    33.4     63   178   73.5   104.    13.1  7.98       1
 8    25.5     74   186.  71.8   130.    14.6  5.93       3
 9    26.8     78   185   75     124.    12.8  8.72       3
10    28.3     85   179   71.3   119.    14.1  6.48       1
# ℹ 142 more rows</code></pre>
</div>
</div>
<p>We can aggregate the data based on groups</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>dat.ID91.summary <span class="ot">&lt;-</span> dat.ID91 <span class="sc">%&gt;%</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">amylase=</span><span class="fu">mean</span>(amylase, <span class="at">na.rm=</span><span class="cn">TRUE</span>),</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">diapow=</span><span class="fu">mean</span>(diapow, <span class="at">na.rm=</span><span class="cn">TRUE</span>),</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">hddate=</span><span class="fu">mean</span>(hddate, <span class="at">na.rm=</span><span class="cn">TRUE</span>),</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">malt=</span><span class="fu">mean</span>(malt, <span class="at">na.rm=</span><span class="cn">TRUE</span>),</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">height=</span><span class="fu">mean</span>(height, <span class="at">na.rm=</span><span class="cn">TRUE</span>),</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">protein=</span><span class="fu">mean</span>(protein, <span class="at">na.rm=</span><span class="cn">TRUE</span>),</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">yield=</span><span class="fu">mean</span>(yield, <span class="at">na.rm=</span><span class="cn">TRUE</span>),</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">Nobs=</span><span class="fu">n</span>()</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>dat.ID91.summary</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 9
  cluster amylase diapow hddate  malt height protein yield  Nobs
    &lt;int&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt;
1       1    26.6   75.7   179.  74.0   110.    13.3  7.71    86
2       2    33.9  110.    179.  74.6   115.    14.8  6.24    17
3       3    28.6   89.9   184.  74.9   121.    13.5  7.56    49</code></pre>
</div>
</div>
<p>The table above shows the summary of mean of each variable based on the cluster. This example illustrates how clustering into groups can be done in R.</p>
</section>
</section>
<section id="k-means-clustering" class="level1">
<h1><span style="color: #234F1E;"><strong>K-Means Clustering</strong></span></h1>
<p>Data clustering techniques serve as descriptive tools for analyzing multivariate datasets and revealing underlying structures within the data. They prove particularly beneficial when traditional second-order statistical methods, such as sample mean and covariance, are not applicable. In exploratory data analysis, where prior knowledge about the dataset’s distribution is lacking, clustering becomes invaluable. It operates as a form of unsupervised classification, forming clusters by assessing similarities and dissimilarities among intrinsic characteristics of different cases. Groupings are based solely on these emergent similarities, independent of any external criteria.</p>
<p>K-Means clustering offers a solution for analyzing unlabeled data, where clear categories or groups are absent. This algorithm provides a straightforward approach to organizing a dataset into a specified number of clusters. It is a simple approach for partitioning a dataset into K distinct, non-overlapping clusters. Through iterative steps, it assigns each data point to one of K groups, relying on feature similarities to cluster the data effectively. The K-Means technique stands out as the most straightforward and widely used clustering approach. It possesses the capability to efficiently group extensive datasets, accomplishing this task with rapid and effective computational speed.</p>
<p>K-means clustering follows similar intuition to hierarchical clustering algorithms used for partitioning the multivariate data set. However, the number of <code>k</code> groups is pre-defined before analysis, hence the name k-means. It allows us to classify observations in a data set in multiple groups called clusters such that observations in the same cluster are as similar as possible <code>(i.e. high intra-class similarity)</code> and observations from other clusters are as dissimilar as possible <code>(i.e. low inter-class similarity)</code>. Each cluster is represented by its center (i.e, centroid) corresponding to the mean of the points assigned to the cluster.</p>
<section id="k-means-clustering-steps" class="level2">
<h2 class="anchored" data-anchor-id="k-means-clustering-steps"><span style="color: #002D62;">K-Means Clustering Steps</span></h2>
<p>Conceptually, K-means operates in the following manner:</p>
<pre><code>1. It randomly selects K centroids.
2. Each data point (e.g., each yield) is assigned to the closest centroid in an n-dimensional space, where n represents the number of features used in clustering (e.g., features such as water, protein, oil, and height). This allocation results in each point belonging to a specific group.
3. The centroids are recalculated as the mean point (vector) of all other points within the group.
4. Steps 2 and 3 are iteratively repeated until either the groups stabilize, meaning no points are reassigned to another centroid, or the maximum number of iterations is reached (commonly set to 10 in default settings).</code></pre>
<p>As you increase the chosen value of K in clustering, the variance within the groups decreases. When K equals the number of observations, each point becomes its own group, resulting in a variance of 0.</p>
<p>The function <code>kmeans()</code> in the base R stats package can be used to perform K-means clustering. It is often recommended to use the <code>set.seed()</code> function to set a seed for <code>R’s</code> random number to make the results reproducible.</p>
<p>The basic argument of the function is as follows:</p>
<ul>
<li><p><code>kmeans(df, k, nstart = x)</code></p>
<ul>
<li><code>df</code> is the numeric data frame</li>
<li><code>k</code> is the number of clusters</li>
<li><code>nstart</code> =<code>x</code> is the number of random starting partitions, where <code>x&gt;1</code> is recommended to have a stable result However, a beautiful cluster visualization can be created from the clusters generated with <code>kmeans()</code> function with <code>fviz_cluster()</code> function in the <code>factoextra</code> package which can be installed using <code>install.packages("factoextra")</code>.</li>
</ul></li>
</ul>
<p>The basic argument needed in the function are:</p>
<ul>
<li><p><code>fviz_cluster(km, df)</code></p>
<ul>
<li><code>km</code> is the kmeans object results</li>
<li><code>df</code> is the original data sets</li>
</ul></li>
</ul>
</section>
<section id="example-1" class="level2">
<h2 class="anchored" data-anchor-id="example-1"><span style="color: #002D62;">Example</span></h2>
<p>To demonstrate how these functions work, let’s consider <code>steptoe.morex.pheno.csv</code> dataset, a multi-environment trial of barley using one environment ,<code>ID91</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"steptoe.morex.pheno.csv"</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>dat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,432 × 10
   gen     env   amylase diapow hddate lodging  malt height protein yield
   &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
 1 Steptoe MN92     22.7     46   150.      NA  73.6   84.5    10.5  5.53
 2 Steptoe MTi92    30.1     72   178       10  76.5   NA      11.2  8.64
 3 Steptoe MTd92    26.7     78   165       15  74.5   75.5    13.4  5.90
 4 Steptoe ID91     26.2     74   179       NA  74.1  111      12.1  8.63
 5 Steptoe OR91     19.6     62   191       NA  71.5   90      11.7  5.34
 6 Steptoe WA91     23.6     54   181       NA  73.8  112      10    6.27
 7 Steptoe MTi91    21       62   181       NA  70.8   98      12    4.10
 8 Steptoe MTd91    NA       NA   181       NA  NA     82      NA    7.07
 9 Steptoe NY92     NA       NA   176        0  NA     77.5    NA    6.05
10 Steptoe ON92     NA       NA   198       50  NA     95      NA    3.70
# ℹ 2,422 more rows</code></pre>
</div>
</div>
<p>Filter the data set to include only the quantitative variables and the genotype name.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>dat.ID91 <span class="ot">&lt;-</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(env<span class="sc">==</span><span class="st">"ID91"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(env, lodging)) <span class="sc">%&gt;%</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"gen"</span>)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat.ID91, <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        amylase diapow hddate malt height protein yield
Steptoe    26.2     74  179.0 74.1  111.0    12.1 8.629
Morex      36.2     97  180.0 76.7  116.0    14.8 7.951
SM1        32.8     83  183.0 74.4  116.8    14.3 6.095
SM2        24.9     81  178.5 72.3  113.0    14.0 7.201</code></pre>
</div>
</div>
<p><strong>To Scale and compute the Euclidean distance</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>dat.ID91.sc <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">scale</span>(dat.ID91), <span class="at">method =</span> <span class="st">"euclidean"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This has scaled the dataset using the euclidean distance</p>
<p><strong>To Compute K-means</strong> <code>(Assuming 3 clusters)</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">234</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>k.means.ID91 <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(dat.ID91.sc, <span class="dv">5</span>, <span class="at">nstart=</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>To Compute Cluster number for each of the observations</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(k.means.ID91<span class="sc">$</span>cluster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Steptoe   Morex     SM1     SM2     SM3     SM4 
      3       5       4       3       3       4 </code></pre>
</div>
</div>
<p>This shows the cluster number for each observations.</p>
<p><strong>To Compute Cluster size</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>k.means.ID91<span class="sc">$</span>size</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25 17 52 39 19</code></pre>
</div>
</div>
<p><strong>To Compute Cluster means</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>k.means.ID91<span class="sc">$</span>centers</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>clustering <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(dat.ID91.sc, <span class="at">centers =</span> <span class="dv">2</span>, <span class="at">nstart =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Using 3 group ( k = 3) we had 41.4% of well-grouped data.</p>
<p>Lets increase the value of k to 4 and compare the results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>clustering <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(dat.ID91.sc, <span class="at">centers =</span> <span class="dv">4</span>, <span class="at">nstart =</span> <span class="dv">10</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>clustering</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With an increase in groups (k = 4), the value of well grouped data increases to 48.4%, which is a better value than the previous. This indicates that the bigger the value of K, the better the grouping gets.</p>
</section>
<section id="k-means-clustering-visualization" class="level2">
<h2 class="anchored" data-anchor-id="k-means-clustering-visualization"><span style="color: #002D62;">K-Means Clustering Visualization</span></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(factoextra)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_cluster</span>(k.means.ID91, <span class="at">data =</span> dat.ID91.sc,</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">ellipse.type =</span> <span class="st">"euclid"</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">labelsize =</span> <span class="dv">8</span>, <span class="at">star.plot=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Multivariate-Analysis_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The objective of clustering analysis is to uncover patterns within the dataset. As illustrated in the plot, observations within a group typically share similar characteristics.</p>
</section>
</section>
<section id="principal-component-analysis-pca" class="level1">
<h1><span style="color: #234F1E;"><strong>Principal Component Analysis (PCA)</strong></span></h1>
<p>Principal Component Analysis (PCA), is a statistical technique aimed at revealing the covariance structure within a set of variables. It facilitates the identification of the primary directions along which the data exhibits variation.</p>
<p>Principal component analysis (PCA) is a useful technique for exploratory data analysis, allowing to better visualize the variation present in a data set with many variables. It is particularly useful in the case of large data sets, where there are many variables; difficult to represent all the data in their raw format, which makes it difficult to understand the trends present inside. PCA allows us to see the general <code>shape</code> of a data, identifying which observations are similar. This can allow us to identify groups of similar observations and determine which variables make one group different from another</p>
<p>The basics of PCA:</p>
<ul>
<li>take a dataset with many variables</li>
<li>reduce the number of variables to a smaller number of <code>principal components</code></li>
<li>visualize the correlations between the variables</li>
<li>visualize the similarities between observations (individuals)</li>
</ul>
<p>The principal components are the <code>directions</code> where there is the most variance, the directions where the data is the most distributed. This means that we are trying to find the straight line that best distributes the data as it is projected along it. The first main component is the straight line which shows the largest variation in the data</p>
<p>PCA is a type of linear transformation that adapts a dataset to a new coordinate system such that the most significant variance is on the first coordinate, and each subsequent coordinate is orthogonal to the last and has a variance lesser. In this way, we transform a set of <code>x</code> correlated variables on <code>y</code> observations into a set of <code>p</code> uncorrelated principal components on the same observations. When many variables are correlated, they will all contribute strongly to the same principal component. Each principal component summarizes a certain percentage of the total variation in the data set. When the initial variables are highly correlated with each other, you will be able to approach most of the complexity of your dataset with just a few main components.</p>
<p>Principal Component Analysis (PCA) is a widely used statistical technique for dimensionality reduction and data visualization. It aims to simplify complex datasets by transforming them into a new coordinate system, where the variables are uncorrelated, and the first few dimensions (principal components) capture the maximum variance in the data.</p>
<section id="heres-how-pca-typically-works" class="level2">
<h2 class="anchored" data-anchor-id="heres-how-pca-typically-works"><span style="color: #002D62;">Here’s how PCA typically works</span></h2>
<ol type="1">
<li><p><strong>Standardization</strong>: If the variables in the dataset are measured on different scales, PCA often starts with standardizing the data to have a mean of 0 and a standard deviation of 1.</p></li>
<li><p><strong>Covariance Matrix</strong>: PCA calculates the covariance matrix of the standardized data. This matrix summarizes the relationships between pairs of variables, indicating how much they vary together.</p></li>
<li><p><strong>Eigenvalue Decomposition</strong>: PCA then decomposes the covariance matrix into its eigenvectors and eigenvalues. The eigenvectors represent the directions of maximum variance in the data, and the corresponding eigenvalues indicate the magnitude of variance along those directions.</p></li>
<li><p><strong>Selecting Principal Components</strong>: The eigenvectors are ranked based on their corresponding eigenvalues, with the first eigenvector capturing the most variance, the second capturing the second most variance, and so on. Typically, only a subset of the top-ranked eigenvectors, called principal components, are retained, thereby reducing the dimensionality of the dataset.</p></li>
<li><p><strong>Dimensionality Reduction</strong>: By projecting the original data onto the selected principal components, PCA creates a lower-dimensional representation of the dataset while preserving as much of the variance as possible. This reduced representation can often capture the essential patterns and structures in the data.</p></li>
</ol>
</section>
<section id="application-of-pca" class="level2">
<h2 class="anchored" data-anchor-id="application-of-pca"><span style="color: #002D62;">Application of PCA</span></h2>
<ul>
<li><p><strong>Dimensionality Reduction</strong>: PCA can reduce the number of variables in a dataset while retaining most of the variance, making it easier to analyze and visualize high-dimensional data.</p></li>
<li><p><strong>Data Visualization</strong>: PCA can be used to visualize high-dimensional data in lower dimensions (e.g., 2D or 3D), allowing for easier interpretation and exploration of data patterns.</p></li>
<li><p><strong>Feature Extraction</strong>: PCA can extract new features (principal components) that are linear combinations of the original variables. These new features may capture important underlying patterns in the data.</p></li>
</ul>
<p>Overall, PCA is a versatile tool widely used in various fields such as statistics, machine learning, finance, image processing, and genetics for data preprocessing, visualization, and feature extraction.</p>
</section>
<section id="example-2" class="level2">
<h2 class="anchored" data-anchor-id="example-2"><span style="color: #002D62;">Example</span></h2>
<p>Let’s consider the <code>australia.soybean.csv</code>, a multi-environment trial of 58 varieties of soybeans, in 4 locations across 2 years in Australia. The description of the data are:</p>
<ul>
<li><strong>env</strong> (environment), 8 levels,</li>
<li><strong>loc</strong> (location)</li>
<li><strong>year</strong></li>
<li><strong>gen</strong> (genotype) of soybeans,</li>
<li><strong>yield</strong>, metric tons/hectare</li>
<li><strong>height</strong>, in meters</li>
<li><strong>lodging</strong>, number of plants</li>
<li><strong>size seed</strong>, in millimetres</li>
<li><strong>protein</strong>, percentage</li>
<li><strong>oil</strong>, percentage</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"australia.soybean.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To study the relationship between two variables, say <code>yield</code> and <code>oil</code>, we can use + Pearson correlation coefficient + Simple linear regression</p>
<p><img src="images/pca-oil-yield.png" class="img-fluid"> The diagram above indicates a positive linear relationship between yield and oil content of soybeans.</p>
<p>Lets consider one location <code>Brookstead</code> in the <code>australia.soybean</code> data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>dat.Brookstead <span class="ot">&lt;-</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(loc<span class="sc">==</span><span class="st">"Brookstead"</span>)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(dat.Brookstead)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>spc_tbl_ [116 × 10] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ env    : chr [1:116] "B70" "B70" "B70" "B70" ...
 $ loc    : chr [1:116] "Brookstead" "Brookstead" "Brookstead" "Brookstead" ...
 $ year   : num [1:116] 1970 1970 1970 1970 1970 1970 1970 1970 1970 1970 ...
 $ gen    : chr [1:116] "G01" "G02" "G03" "G04" ...
 $ yield  : num [1:116] 1.253 1.167 0.468 1.445 1.338 ...
 $ height : num [1:116] 1.01 1.13 1.16 1.24 1.12 ...
 $ lodging: num [1:116] 3.25 2.75 2.25 1.5 2 2.25 2 2.25 1.75 2 ...
 $ size   : num [1:116] 8.85 8.9 10.8 10.6 11.95 ...
 $ protein: num [1:116] 39.5 38.6 37.8 38.7 37.8 ...
 $ oil    : num [1:116] 18.9 19.8 20.4 20.4 20.8 ...
 - attr(*, "spec")=
  .. cols(
  ..   env = col_character(),
  ..   loc = col_character(),
  ..   year = col_double(),
  ..   gen = col_character(),
  ..   yield = col_double(),
  ..   height = col_double(),
  ..   lodging = col_double(),
  ..   size = col_double(),
  ..   protein = col_double(),
  ..   oil = col_double()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
</div>
<p>We store the names of the genotypes in a vector</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>labels.gen <span class="ot">&lt;-</span> dat.Brookstead<span class="sc">$</span>gen</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And only consider the quantitative variables</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>dat.Brookstead <span class="ot">&lt;-</span> dat.Brookstead <span class="sc">%&gt;%</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(env, loc, year, gen))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We run the PCA using the prcomp function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>dat.Brookstead.pca<span class="ot">&lt;-</span><span class="fu">prcomp</span>(dat.Brookstead, <span class="at">center=</span><span class="cn">TRUE</span>, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(dat.Brookstead.pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Importance of components:
                          PC1    PC2    PC3     PC4     PC5     PC6
Standard deviation     1.8780 0.9949 0.9176 0.60462 0.42702 0.30580
Proportion of Variance 0.5878 0.1650 0.1403 0.06093 0.03039 0.01559
Cumulative Proportion  0.5878 0.7528 0.8931 0.95402 0.98441 1.00000</code></pre>
</div>
</div>
<p>The first 2 axes explain more than 75% of the variation in the dataset. We visualize the PCA using a <code>biplot</code>. A <code>biplot</code> is a type of graph that allows us to simultaneously visualize the observations and the variables. We use the <code>ggbiplot</code> package to visualize the biplot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbiplot)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggbiplot</span>(dat.Brookstead.pca) <span class="sc">+</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Yield and other traits of soybeans in Australia"</span>,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Period 1970-1971"</span>,</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">caption =</span> <span class="st">"Data: agridat::australia.soybean"</span></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/pca-biplot-1.png" width="65%"></p>
<p>The components are seen as arrows coming from the origin. Here, all the variables contribute to PC1, with:</p>
<ul>
<li>higher values of <code>oil</code>, <code>size</code> and <code>yield</code> to the right on this graph and lower values to the left</li>
<li>vice versa for <code>protein</code>, <code>lodging</code> and <code>height</code></li>
</ul>
<p>This allows us to see how the observations relate to the components. However, this is not very informative if we do not identify the observations. We can display the names of the genotypes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbiplot)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggbiplot</span>(dat.Brookstead.pca, <span class="at">labels=</span>labels.gen) <span class="sc">+</span> <span class="co">#&lt;&lt;</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Yield and other traits of soybeans in Australia"</span>,</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Period 1970-1971"</span>,</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">caption =</span> <span class="st">"Data: agridat::australia.soybean"</span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/pca-biplot-2.png" width="65%"></p>
<p>The diagram shows in detail the observations per each genotype.</p>
<p>We have two years of testing, 1970 and 1971. Could be interesting to identify the genotypes by year</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbiplot)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggbiplot</span>(dat.Brookstead.pca, <span class="at">labels=</span>labels.gen, <span class="at">groups=</span>labels.year)<span class="sc">+</span> <span class="co">#&lt;&lt;</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Yield and other traits of soybeans in Australia"</span>,</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Period 1970-1971"</span>,</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">caption =</span> <span class="st">"Data: agridat::australia.soybean"</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/pca-biplot-3.png" width="65%"></p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>